### ISSUE TITLE
Parameters in TW::AdditionalData not propogated to TW::SolverBase

### ISSUE THREAD

**Issue by jppelteret** [dealii/dealii#4443]
Refers to [this post](https://groups.google.com/forum/#!topic/dealii/kiLFKj94ubc) on the forum.

It looks like the parameters set in the TrilinosWrappers iterative solver AdditionalData structs are not passed to the SolverBase class. Thus the instance of its own AdditionalData only ever stores its default values, and these defaults are later referenced [here](https://github.com/dealii/dealii/blob/master/source/lac/trilinos_solver.cc#L452) and [here](https://github.com/dealii/dealii/blob/master/source/lac/trilinos_solver.cc#L468). 

Proposed patch: Remove all AdditionalData instances stored in the child iterative solver classes and change their constructors to
```c++
  SolverCG::SolverCG (SolverControl        &cn,
                      const AdditionalData &data)
    :
    SolverBase (cn, 
                SolverBase::AdditionalData(data.output_solver_details))
  {
    solver_name = cg;
  }
```
etc.

**Comment by bangerth** [comment]
Passing the respective flag down to the base class is clearly necessary, but if `AdditionalData` contains additional data (in addition to the data stored by the base class), then that needs to be stored by the derived class as well.

In practice, I would think that the `AdditionalData` structures of the derived classes are either identical to the one in the base class, or derived from the base class. In those cases, you can just do
```
SolverCG::SolverCG (SolverControl        &cn,
                      const AdditionalData &data)
    :
    SolverBase (cn, 
                SolverBase::AdditionalData(data))  // automatically cast 'data' to base class
  {
    solver_name = cg;
  }
```

**Comment by jppelteret** [comment]
Yes, I agree with this approach. We don't use inheritance at the moment, so obviously it means that we have to change the current implementation of each solver's `AdditionalData` to derive from the base class' `AdditionalData`. As far as I can see this shouldn't cause any difficulties though.

**Comment by juan42** [comment]
Hi,

i'm back in the office now and working on the fix. Shouldn't
```
static_cast<const SolverBase::AdditionalData&>
```
be preferred over
```
SolverBase::AdditionalData(data)) 
```
for the sake of clarity?

**Comment by juan42** [comment]
After looking at the problem more in depth, i noticed that every AdditionalData struct has multiple instances. One for the solver specific and one for the SolverBase::AdditionalData. This should be cleaned up since i cannot make sense of it why its needed. This can only be resolved iff the solver is created in the subclass (SolverCG for example) and not in the SolverBase with a switch case. Otherwise there would be (actually at the moment there is) a dependency from the base class to the inherited class.

Thoughts?

**Comment by jppelteret** [comment]
Hi @juan42 ,

Sorry again for the delay in getting back to you on this. With regard to the preference of calling the copy constructor directly versus explicitly casting it to the base class, it shouldn't make a difference in practise so either should be fine. If we were to use an inheritance-based approach then one should in fact just be able to pass the `data` object directly to the base class than expects a `SolverBase::AdditionalData`.

On the second point, I think that the idea behind the current structure was likely to hide this additional `gmres_restart_parameter` parameter that's only needed by the one solver. Do you propose changing the [`switch (solver_name)`](https://github.com/dealii/dealii/blob/master/source/lac/trilinos_solver.cc#L442) to a call to a virtual function in SolverBase, e.g.
```c++
virtual void set_solver(AztecOO &solver) = 0;
```
and, for GMRES then having
```c++
virtual void set_solver(AztecOO &solver)
{
  solver.SetAztecOption(AZ_solver, AZ_gmres);
  solver.SetAztecOption(AZ_kspace, additional_data.gmres_restart_parameter);
}
```
? This would could work nicely because then each solver would configure itself, and the additional parameter currently [awkwardly located in SolverBase::AdditionalData](https://github.com/dealii/dealii/blob/master/include/deal.II/lac/trilinos_solver.h#L112) could be held by GMRES' own copy of `additional_data`. The primary disadvantage to this approach is that the [option to simply use the base solver class like a solver selector](https://github.com/dealii/dealii/blob/master/include/deal.II/lac/trilinos_solver.h#L72) instead of instantiating one of the derived classes would disappear. Unless we can devise some way in which to provide backwards compatibility before deprecating this option, this change might be too radical to implement now.

@kronbichler What are your views on the idea of each solver configuring itself?

**Comment by juan42** [comment]
> Unless we can devise some way in which to provide backwards compatibility before deprecating this option, this change might be too radical to implement now.

That is my concern as well. The interface would either change or need another call after the "factory" (right now the SolverBase class) is created.

**Comment by jppelteret** [comment]
Ok, well then lets just keep to the point for now and solve the problem at hand. We can open another issue and work though refining the overall class structure later.

Do you then plan on removing the (unused) `additional_data` instances in the derived solver classes and replacing their `AdditionalData` structs with `using SolverBase::AdditionalData;` to maintain backwards compatibility?

**Comment by juan42** [comment]
> Do you then plan on removing the (unused) additional_data instances in the derived solver classes and replacing their AdditionalData structs with using SolverBase::AdditionalData; to maintain backwards compatibility?

That won't work since https://github.com/dealii/dealii/blob/master/source/lac/trilinos_solver.cc#L452 needs the data in the base struct (which seems to be some circular dependency?).

**Comment by jppelteret** [comment]
No, there's no circular dependency. That line refers to the `additional_data` [stored here](https://github.com/dealii/dealii/blob/master/include/deal.II/lac/trilinos_solver.h#L341) in `SolverBase`.  In addition to this, each solver stores an instance of its "specialised" `AdditionalData` (e.g. [here](https://github.com/dealii/dealii/blob/master/include/deal.II/lac/trilinos_solver.h#L400)) which they **could** call if necessary, but in practise never do (apart from [initialisation](https://github.com/dealii/dealii/blob/master/source/lac/trilinos_solver.cc#L609), of course). 

So I advocate removing all of the redundant instances of the "specialised" `AdditionalData` in `SolverCG`, `SolverGMRES` etc. and just having a single copy of a `SolverBase::AdditionalData` in `SolverBase` itself. To maintain compatibility, we would then need to do the following
```c++
  class SolverCG : public SolverBase
  {
  public:

    using  SolverBase::AdditionalData;

    SolverCG (SolverControl        &cn,
              const AdditionalData &data = AdditionalData())
    :
    SolverBase (cn, data) // No cast necessary since AdditionalData == SolverBase::AdditionalData
  {
    solver_name = cg;
  }
};
```
so that the following user code would still work:
```c++
TrilinosWrappers::SolverCG::AdditionalData data (true);
TrilinosWrappers::SolverCG solver (solver_control, data);
```

**Comment by juan42** [comment]
I tried to implement my fix in PR #4512 . In my test it works fine but i also needed to remove the restart parameter.
